File: app/src/main/java/farasense/mobile/view/ui/activity/fragment/MonthlyChartConsumptionFragment.java
package farasense.mobile.view.ui.activity.fragment;

import android.arch.lifecycle.LifecycleRegistry;
import android.arch.lifecycle.ViewModelProviders;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.github.mikephil.charting.charts.BarChart;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.components.YAxis;
import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.BarDataSet;
import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.highlight.Highlight;

import java.util.List;

import javax.annotation.Nullable;

import farasense.mobile.R;
import farasense.mobile.databinding.MonthlyChartConsumptionFragmentDataBinding;
import farasense.mobile.util.ChartUtil;
import farasense.mobile.view_model.MonthlyChartConsumptionFragmentViewModel;

public class MonthlyChartConsumptionFragment extends Fragment {

    private final LifecycleRegistry registry = new LifecycleRegistry(this);
    private MonthlyChartConsumptionFragmentDataBinding binding;
    private MonthlyChartConsumptionFragmentViewModel viewModel;
    private BarChart monthlyChart;
    private List<BarEntry> entryList;
    private BarDataSet dataSet;
    private BarData data;

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_monthly_chart_consumption, container, false);

        viewModel = ViewModelProviders.of(getActivity()).get(MonthlyChartConsumptionFragmentViewModel.class);

        View view = binding.getRoot();

        // createMonthlyChart();

        binding.setMonthlyCompumptionFragment(viewModel);

        return view;
    }

    private void createMonthlyChart() {
        monthlyChart = binding.monthlyChartConsumption;
        monthlyChart.getDescription().setText(getResources().getString(R.string.description_daily_chart));
        monthlyChart.setScaleYEnabled(false);
        monthlyChart.setFitBars(true);
        monthlyChart.zoom(3.4F, 0, 0,0);

        monthlyChart.moveViewToX(20);

        configureValueSelectedListener(monthlyChart);

        XAxis xAxis = monthlyChart.getXAxis();
        YAxis yAxis = monthlyChart.getAxisLeft();

        entryList = viewModel.getMonthlyConsumption();
        xAxis.setValueFormatter(ChartUtil.setChartLabels(viewModel.getMonthlyChartLabels()));

        configureAxis(monthlyChart, xAxis, yAxis);

        dataSet = configureDataSet(entryList);

        data = new BarData(dataSet);

        monthlyChart.setData(data);
    }

    private void configureAxis(BarChart dailyChart, XAxis xAxis, YAxis yAxis) {
        dailyChart.animateY(3000);

        xAxis.setDrawAxisLine(false);
        xAxis.setDrawGridLines(false);
        xAxis.setEnabled(true);

        dailyChart.getAxisRight().setEnabled(false);
        yAxis.setDrawLabels(true);
        yAxis.setDrawAxisLine(false);
        yAxis.setDrawGridLines(true);
        yAxis.setDrawZeroLine(true);

        xAxis.setPosition(XAxis.XAxisPosition.BOTTOM);
        xAxis.setGranularity(1F);
    }

    private BarDataSet configureDataSet(List<BarEntry> entryList) {
        BarDataSet dataSet;
        dataSet = new BarDataSet(entryList, getResources().getString(R.string.kilowatts));
        dataSet.setColor(getResources().getColor(R.color.colorTabBar));
        dataSet.setDrawValues(false);

        return dataSet;
    }

    private void configureValueSelectedListener(BarChart monthlyChart) {
//        monthlyChart.setOnChartValueSelectedListener(new com.github.mikephil.charting.listener.OnChartValueSelectedListener() {
//            @Override
//            public void onValueSelected(Entry entry, Highlight h) {
//                if (binding.labelDailyValueSelected.getVisibility() == View.VISIBLE) {
//                    binding.labelDailyValueSelected.setText(String.valueOf(entry.getY()));
//                } else {
//                    binding.labelDailyValueSelected.setVisibility(View.VISIBLE);
//                    binding.labelDailyValueSelected.setText(String.valueOf(entry.getY()));
//                }
//            }
//
//            @Override
//            public void onNothingSelected() {
//                if (binding.labelDailyValueSelected.getVisibility() == View.VISIBLE) {
//                    binding.labelDailyValueSelected.setVisibility(View.GONE);
//                } else {
//                    binding.labelDailyValueSelected.setVisibility(View.GONE);
//                }
//            }
//        });
    }

    @Override
    public LifecycleRegistry getLifecycle() { return registry; }

}

File: app/src/main/java/old_farasense/view/ui/activity/fragment/MonthlyChartConsumptionFragment.kt
//package old_farasense.view.ui.activity.fragment
//
//import android.arch.lifecycle.ViewModelProviders
//import android.databinding.DataBindingUtil
//import android.os.Bundle
//import android.support.v4.app.Fragment
//import android.view.LayoutInflater
//import android.view.View
//import android.view.ViewGroup
//import com.github.mikephil.charting.charts.BarChart
//import com.github.mikephil.charting.components.XAxis
//import com.github.mikephil.charting.components.YAxis
//import com.github.mikephil.charting.data.BarData
//import com.github.mikephil.charting.data.BarDataSet
//import com.github.mikephil.charting.data.BarEntry
//import com.github.mikephil.charting.data.Entry
//import com.github.mikephil.charting.highlight.Highlight
//import farasense.mobile.R
//import farasense.mobile.databinding.MonthlyChartConsumptionFragmentDataBinding
//import old_farasense.util.ChartUtil
//import old_farasense.view_model.MonthlyChartConsumptionFragmentViewModel
//
//class MonthlyChartConsumptionFragment : Fragment() {
//
//    private var binding: MonthlyChartConsumptionFragmentDataBinding? = null
//    private var viewModel: MonthlyChartConsumptionFragmentViewModel? = null
//    private var monthlyChart: BarChart? = null
//    private var entryList: List<BarEntry>? = null
//    private var dataSet: BarDataSet? = null
//    private var data: BarData? = null
//
//    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
//        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_monthly_chart_consumption, container, false)
//
//        viewModel = ViewModelProviders.of(activity!!).get(MonthlyChartConsumptionFragmentViewModel::class.java)
//
//        val view = binding?.root
//
//        createMonthlyChart()
//
//        binding?.monthlyCompumptionFragment = viewModel
//
//        return view
//    }
//
//    private fun createMonthlyChart() {
//        monthlyChart = binding?.monthlyChartConsumption
//        monthlyChart?.description?.text = resources.getString(R.string.description_monthly_chart)
//        monthlyChart?.isScaleYEnabled = false
//        monthlyChart?.setFitBars(true)
//        monthlyChart?.zoom(3.0f, 0f, 0f, 0f)
//
//        monthlyChart?.moveViewToX(20f)
//
//        configureValueSelectedListener(monthlyChart!!)
//
//        val xAxis = monthlyChart?.xAxis
//        val yAxis = monthlyChart?.axisLeft
//
//        entryList = viewModel?.monthlyConsumption
//        xAxis?.valueFormatter = viewModel?.getMonthlyChartLabels()?.let { ChartUtil.setChartLabels(it) }
//
//        xAxis?.let { yAxis?.let { it1 -> configureAxis(monthlyChart!!, it, it1) } }
//
//        dataSet = configureDataSet(entryList)
//
//        data = BarData(dataSet!!)
//
//        monthlyChart?.data = data
//    }
//
//    private fun configureAxis(dailyChart: BarChart, xAxis: XAxis, yAxis: YAxis) {
//        dailyChart.animateY(3000)
//
//        xAxis.setDrawAxisLine(false)
//        xAxis.setDrawGridLines(false)
//        xAxis.isEnabled = true
//
//        dailyChart.axisRight.isEnabled = false
//        yAxis.setDrawLabels(true)
//        yAxis.setDrawAxisLine(false)
//        yAxis.setDrawGridLines(true)
//        yAxis.setDrawZeroLine(true)
//
//        xAxis.position = XAxis.XAxisPosition.BOTTOM
//        xAxis.granularity = 1f
//    }
//
//    private fun configureDataSet(entryList: List<BarEntry>?): BarDataSet {
//        val dataSet: BarDataSet
//        dataSet = BarDataSet(entryList!!, resources.getString(R.string.kilowatts))
//        dataSet.color = resources.getColor(R.color.colorMonthlyChartBar)
//        dataSet.setDrawValues(false)
//
//        return dataSet
//    }
//
//    private fun configureValueSelectedListener(monthlyChart: BarChart) {
//        monthlyChart.setOnChartValueSelectedListener(object : com.github.mikephil.charting.listener.OnChartValueSelectedListener {
//            override fun onValueSelected(entry: Entry, h: Highlight) {
//                if (binding?.labelMonthlyValueSelected?.visibility == View.VISIBLE) {
//                    binding?.labelMonthlyValueSelected?.text = entry.y.toString()
//                } else {
//                    binding?.labelMonthlyValueSelected?.visibility = View.VISIBLE
//                    binding?.labelMonthlyValueSelected?.text = entry.y.toString()
//                }
//            }
//
//            override fun onNothingSelected() {
//                if (binding?.labelMonthlyValueSelected?.visibility == View.VISIBLE) {
//                    binding?.labelMonthlyValueSelected?.visibility = View.GONE
//                } else {
//                    binding?.labelMonthlyValueSelected?.visibility = View.GONE
//                }
//            }
//        })
//    }
//}
